File: frontend/lib/controllers/network_controller.dart
import 'dart:convert';
import 'dart:io';

import 'package:camera/camera.dart';
import 'package:frontend/utils/config.dart';
import 'package:frontend/utils/response.dart';

import 'package:http/http.dart' as http;

class NetworkController {
  final String _url = cUrl;

  login(String email, String password) async {
    final httpResponse = await http.post(
      Uri.parse("$_url/sign-in"),
      body: {
        'email': email,
        'password': password,
      },
    );
    final jsonResponse = await jsonDecode(httpResponse.body);
    return Response(jsonResponse["type"], jsonResponse["data"]);
  }

  register(String name, String email, String passwordHash, XFile image) async {
    final request = http.MultipartRequest("POST", Uri.parse("$_url/sign-up"));
    final headers = {"Content-type": "multipart/form-data"};
    request.fields['name'] = name;
    request.fields['email'] = email;
    request.fields['password'] = passwordHash;
    File imageFile = File(image.path);
    request.files.add(http.MultipartFile(
        'image', image.readAsBytes().asStream(), imageFile.lengthSync(),
        filename: imageFile.path.split("/").last));
    request.headers.addAll(headers);
    http.Response httpResponse =
        await http.Response.fromStream(await request.send());
    final jsonResponse = await jsonDecode(httpResponse.body);
    return Response(jsonResponse["type"], jsonResponse["data"]);
  }

  checkIn(XFile image) async {
    final request = http.MultipartRequest("POST", Uri.parse("$_url/check-in"));
    return await markAttendance(image, request);
  }

  checkOut(XFile image) async {
    final request = http.MultipartRequest("POST", Uri.parse("$_url/check-out"));
    return await markAttendance(image, request);
  }

  markAttendance(XFile xFileImage, http.MultipartRequest request) async {
    File image = File(xFileImage.path);
    final headers = {"Content-type": "multipart/form-data"};
    request.files.add(http.MultipartFile(
        'image', image.readAsBytes().asStream(), image.lengthSync(),
        filename: image.path.split("/").last));
    request.headers.addAll(headers);
    http.Response httpResponse =
        await http.Response.fromStream(await request.send());
    final jsonResponse = await jsonDecode(httpResponse.body);
    return Response(jsonResponse["type"], jsonResponse["data"]);
  }

  getAttendance(int userId) async {
    final request = http.MultipartRequest("GET", Uri.parse("$_url/attendance"));
    request.fields['id'] = userId.toString();
    http.Response httpResponse =
        await http.Response.fromStream(await request.send());
    final jsonResponse = await jsonDecode(httpResponse.body);
    return jsonResponse["data"];
  }

  getUsers() async {
    final request = http.MultipartRequest("GET", Uri.parse("$_url/users"));
    http.Response httpResponse =
        await http.Response.fromStream(await request.send());
    final jsonResponse = await jsonDecode(httpResponse.body);
    return jsonResponse["data"];
  }

  deleteEmployee(int id) async {
    await http.post(
      Uri.parse("$_url/delete-employee"),
      body: {'id': id.toString()},
    );
  }
}

File: backend/controller/network_controller.py
from flask import Flask, request, jsonify

from backend.service.service import Service


class NetworkController:

    def __init__(self, ip, port, service: Service):
        self.ip = ip
        self.port = port
        self.service = service
        self.app = Flask(__name__)

    def run(self):
        @self.app.route('/sign-up', methods=['POST'])
        def sign_up():
            if request.method == "POST":
                name = request.form['name']
                email = request.form['email']
                password = request.form['password']
                image_file = request.files['image']
                print(name, email, password)
                response = self.service.sign_up(name, email, password, image_file)
                return jsonify({
                    "type": response.type,
                    "data": None
                })

        @self.app.route('/sign-in', methods=['POST'])
        def sign_in():
            if request.method == "POST":
                email = request.form['email']
                password = request.form['password']
                response = self.service.sign_in(email, password)
                return jsonify({
                    "type": response.type,
                    "data": str(response.data)
                })

        @self.app.route('/check-in', methods=['POST'])
        def check_in():
            if request.method == "POST":
                image_file = request.files['image']
                response = self.service.check_in(image_file)
                return jsonify({
                    "type": response.type,
                    "data": str(response.data)
                })

        @self.app.route('/check-out', methods=['POST'])
        def check_out():
            if request.method == "POST":
                imageFile = request.files['image']
                response = self.service.check_out(imageFile)
                return jsonify({
                    "type": response.type,
                    "data": str(response.data)
                })

        @self.app.route('/attendance', methods=['GET'])
        def attendance():
            if request.method == "GET":
                id_ = int(request.form['id'])
                response = self.service.get_user_attendance(id_)
                return jsonify({
                    "data": response
                })

        @self.app.route('/users', methods=['GET'])
        def users():
            if request.method == "GET":
                response = self.service.get_all_non_admin_users()
                return jsonify({
                    "data": response
                })

        @self.app.route('/delete-employee', methods=['POST'])
        def delete_employee():
            if request.method == "POST":
                id_ = int(request.form['id'])
                self.service.delete_employee(id_)
            return None


        self.app.run(debug=True, host=self.ip, port=self.port)
    